docs/reuse/best-practices.txt

```diff
@@ -0,0 +1,36 @@
+- Use absolute paths in subprocesses to prevent security issues.
+- Execute processes directly rather than via the shell.
+
+**[How to write and structure charm code](https://documentation.ubuntu.com/ops/latest/howto/write-and-structure-charm-code/)**
+- Name the repository using the pattern ``<charm name>-operator`` for a single charm, or ``<base charm name>-operators`` when the repository will hold multiple related charms. For the charm name, see {external+charmcraft:ref}`Charmcraft | Specify a name <specify-a-name>`. See: [Create a repository and initialise it](#create-a-repository-and-initialise-it).
+- Set the [`requires-python`](https://packaging.python.org/en/latest/specifications/pyproject-toml/#requires-python) version in your `pyproject.toml` so that tooling will detect any use of Python features not available in the versions you support. See: [Use the Python provided by the base](#define-the-required-dependencies).
+- Ensure that tooling is configured to automatically detect new versions, particularly security releases, for all your dependencies. See: [Add Python dependencies to pyproject.toml and generate a lock file](#define-the-required-dependencies).
+- Ensure that the `pyproject.toml` *and* the lock file are committed to version control, so that exact versions of charms can be reproduced. See: [Add Python dependencies to pyproject.toml and generate a lock file](#define-the-required-dependencies).
+- The quality assurance pipeline of a charm should be automated using a continuous integration (CI) system. See: [Validate your charm with every change](#validate-your-charm-with-every-change).
+
+**[How to log from your charm](https://documentation.ubuntu.com/ops/latest/howto/log-from-your-charm/)**
+- Capture output to `stdout` and `stderr` in your charm and use the logging and warning functionality to send messages to the charm user, rather than rely on Juju capturing output. In particular, you should avoid `print()` calls, and ensure that any subprocess calls capture output.
+- Ensure that log messages are clear, meaningful, and provide enough information for the user to troubleshoot any issues. Avoid spurious logging. For instance, try not to log when event handlers are called, as the Juju controller does this automatically.
+-Never log credentials or other sensitive information.
+
+**[How to manage charms](https://documentation.ubuntu.com/ops/latest/howto/manage-charms/)**
+- All charms should provide the commands configured by the Charmcraft profile, to allow easy testing across the charm ecosystem. It's fine to tweak the configuration of individual tools, or to add additional commands, but keep the command names and meanings that the profile provides. See: [Develop your charm](#develop-your-charm).
+- The quality assurance pipeline of a charm should be automated using a continuous integration (CI) system. See: [Develop your charm](#develop-your-charm).
```

Comment:
In the charmhub-listing-review code I have this skipped since it's a dupe of the one above. I could add a skip here as well.

It's probably actually better if we drop one of them - either totally, or making one just regular inline text or a tip.